### PR TITLE
Allow trailing commas in function parameter and argument lists

### LINKED ISSUE #7279: Trailing Commas in Function Param Lists
https://jeffmo.github.io/es-trailing-function-commas/

This proposal is Stage 3 according to https://github.com/tc39/ecma262, so seems like its at an appropriate stability for TypeScript team to implement it.

This would allow function definitions and calls in the following style: 

```ts
function foo(
  bar: Bar, // Comment about bar
  baz: Baz, // Comment about baz
) {
  // Implementation...
}

foo(
  bar,
  baz,
);
```

Benefits: https://github.com/jeffmo/es-trailing-function-commas/blob/master/proposal_presentation_slides.pdf

 * Helps with VCS authorship attribution in case of added/removed parameters
 * Consistency with object and array literals
 * Lintable (if you don't like the style, can ban it in your codebase with a linter)
 * Supported by other languages

In addition to all of the above, I suspect the extra "verbosity" of TypeScript's inline types makes this feature even more attractive in TypeScript than in vanilla JS. I feel this particularly in class constructors, which can also have `private` annotations, and possibly a `readonly` annotation coming down the pipe.

```ts
class Foo {
  constructor(
    private bar: Bar = new BasicBar(),
    private baz: Baz = new BasicBaz(),
  ) {}
}
```

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #7279

If this is pulled we should close #7747 too.

### CODE REVIEW

**Comment by mhegazy** [comment]
can you also add a test for trailing commas after a rest arg.

**Comment by mhegazy** [comment]
:+1:

**Comment by andy-ms** [comment]
Need review on 2fc2f5c

**Comment by mhegazy** [comment]
looks good.

**Comment by mhegazy** [comment]
i would also add a test for getters, setters, call and construct signatures.

File: tests/baselines/reference/trailingCommasInFunctionParametersAndArguments.errors.txt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+tests/cases/conformance/es7/trailingCommasInFunctionParametersAndArguments.ts(20,11): error TS1138: Parameter declaration expected.
+
+
+==== tests/cases/conformance/es7/trailingCommasInFunctionParametersAndArguments.ts (1 errors) ====
```
Comment (sandersn) [inline tests/baselines/reference/trailingCommasInFunctionParametersAndArguments.errors.txt]: that's weird, I wasn't expecting to see a .errors.txt and a .types. Usually they are mutually exclusive. Did the .types file manage to skip the broken getter/setter pair?

**Comment by sandersn** [comment]
:+1:

**Pull request merged**